src/amo/components/SearchForm/index.js

```diff
@@ -0,0 +1,236 @@
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import { compose } from 'redux';
+import Autosuggest from 'react-autosuggest';
+import { withRouter } from 'react-router';
+import defaultDebounce from 'simple-debounce';
+
+import {
+  ADDON_TYPE_EXTENSION,
+  ADDON_TYPE_THEME,
+  validAddonTypes,
+} from 'core/constants';
+import log from 'core/logger';
+import translate from 'core/i18n/translate';
+import { convertFiltersToQueryParams } from 'core/searchUtils';
+import SearchInput from 'ui/components/SearchInput';
+import LoadingText from 'ui/components/LoadingText';
+import { withErrorHandler } from 'core/errorHandler';
+import {
+  autocompleteCancel,
+  autocompleteStart,
+} from 'core/reducers/autocomplete';
+import Suggestion from 'amo/components/SearchForm/Suggestion';
+
+import 'core/css/inc/lib.scss';
+import './styles.scss';
+
+
+export class SearchFormBase extends React.Component {
+  static propTypes = {
+    addonType: PropTypes.string,
+    api: PropTypes.object.isRequired,
+    i18n: PropTypes.object.isRequired,
+    pathname: PropTypes.string.isRequired,
+    query: PropTypes.string.isRequired,
+    dispatch: PropTypes.func.isRequired,
+    errorHandler: PropTypes.object.isRequired,
+    suggestions: PropTypes.arrayOf(PropTypes.shape({
+      name: PropTypes.string.isRequired,
+      url: PropTypes.string.isRequired,
+      iconUrl: PropTypes.string.isRequired,
+    })).isRequired,
+    loadingSuggestions: PropTypes.bool.isRequired,
+    router: PropTypes.object.isRequired,
+    debounce: PropTypes.func.isRequired,
+  }
+
+  static defaultProps = {
+    debounce: defaultDebounce,
+  };
+
+  constructor(props: Object) {
+    super(props);
+
+    this.state = {
+      searchValue: props.query || '',
+    };
+
+    this.handleSuggestionsFetchRequested = props.debounce(
+      this.handleSuggestionsFetchRequested.bind(this),
+      200
+    );
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const { query } = nextProps;
+
+    if (this.props.query !== query) {
+      this.setState({ searchValue: query || '' });
+    }
+  }
+
+  getSuggestions() {
+    if (this.props.loadingSuggestions) {
+      // 10 is the maximum number of results returned by the API
+      return Array(10).fill({ component: <LoadingText width={60} /> });
+    }
+
+    return this.props.suggestions;
+  }
+
+  goToSearch(query) {
+    const { addonType, api, pathname, router } = this.props;
+    const filters = { query };
+
+    if (addonType) {
+      filters.addonType = addonType;
+    }
+
+    router.push({
+      pathname: `/${api.lang}/${api.clientApp}${pathname}`,
+      query: convertFiltersToQueryParams(filters),
+    });
+  }
+
+  handleSearch = (e) => {
+    e.preventDefault();
+    this.searchInput.blur();
+    this.goToSearch(this.searchInput.value);
+  }
+
+  handleSearchChange = (e) => {
+    this.setState({ searchValue: e.target.value });
+  }
+
+  handleSuggestionsFetchRequested({ value }) {
```

Comment:
I find that snippet in the react docs (`this.handleClick = this.handleClick.bind(this)`) to be a horrendous hack :/

The main problem I have with that approach is just that there's no way to look at the method and *know* that it has been replaced. You have to also think to look in the constructor which should not be a requirement for figuring out how things work. 

If you were to do it in the constructor using a *new* property then I'd be fine with that since one would know to search for the property by name. In other words, I just don't want to see a method by the same name get replaced.